caffe2/core/nomnigraph/include/nomnigraph/Graph/Algorithms.h

```diff
@@ -26,6 +26,87 @@ template <typename T, typename U>
 std::vector<Subgraph<T, U>> tarjans(Graph<T, U> *g);
 #include "nomnigraph/Graph/TarjansImpl.h"
 
+
+/// \brief A binary graph matching algorithm based on Kahn's algorithm.
+template <typename T, typename U, typename F>
+std::vector<Subgraph<T, U>> binaryMatch(Graph<T, U> *g, F condition) {
+  using G = Graph<T, U>;
+  auto edges = g->getMutableEdges();
+  std::unordered_set<typename G::EdgeRef> edgeSet(edges.begin(), edges.end());
+  // Topologically sorted nodes.
+  std::vector<Subgraph<T,U>> sortedNodes;
+  sortedNodes.emplace_back();
+
+  // Find the initial frontier.
+  std::vector<typename G::NodeRef> frontier;
+  std::vector<typename G::NodeRef> nextFrontier;
+
+  for (auto n : g->getMutableNodes()) {
+    if (n->getInEdges().size() == 0 && condition(n)) {
```

Comment:
If no primary input nodes matches condition, the algorithm will fail to start. We probably need to do one iteration to jump start in this case.

Reply:
ah good catch, I think it'd be as simple as adding the inverse of the condition to the nextFrontier and then refactoring the while loop

Reply:
Yes.